src/jailhouse/jailhouse_driver.c

```diff
@@ -1,35 +1,184 @@
+    if (virJailhouseConfigInitialize() < 0)
+        return NULL;
+
+    if (!(cfg = virObjectNew(virJailhouseDriverConfigClass)))
+        return NULL;
+
+    cfg->sys_config_file_path = g_strdup(DATADIR "/jailhouse/system.cell");
+
+    cfg->cell_config_dir = g_strdup(DATADIR "/jailhouse/cells");
+
+    return cfg;
+}
+
+static void virJailhouseDriverConfigDispose(void *obj)
+{
+
+    virJailhouseDriverConfigPtr cfg = obj;
+
+    VIR_FREE(cfg->sys_config_file_path);
+    VIR_FREE(cfg->cell_config_dir);
+}
+
+static int
+jailhouseLoadConf(virJailhouseDriverConfigPtr config)
+{
+    g_autoptr(virConf) conf = NULL;
+
+    if (!(conf = virConfReadFile(JAILHOUSE_CONFIG_FILE, 0)))
+        return -1;
+
+    if (virConfGetValueString(conf, "system_config",
+                              &config->sys_config_file_path) < 0)
+        return -1;
+
+    if (virConfGetValueString(conf, "non_root_cells_dir",
+                              &config->cell_config_dir) < 0)
+        return -1;
+
+    return 0;
+}
+
+static int
+jailhouseCreateAndLoadCells(virJailhouseDriverPtr driver)
+{
+    if (!driver->config ||
+        !driver->config->cell_config_dir ||
+        strlen(driver->config->cell_config_dir) == 0)
+        return -1;
+
+    // Create all cells in the hypervisor.
+    if (createJailhouseCells(driver->config->cell_config_dir) < 0)
+        return -1;
+
+    // Get all cells created above.
+    driver->cell_info_list = getJailhouseCellsInfo();
+
+    return 0;
+}
+
+static void
+jailhouseFreeDriver(virJailhouseDriverPtr driver)
+{
+    if (!driver)
+        return;
+
+    VIR_FREE(driver->config);
+    VIR_FREE(driver);
+}
 
 static virDrvOpenStatus
 jailhouseConnectOpen(virConnectPtr conn,
-                     virConnectAuthPtr auth,
-                     virConfPtr conf,
-                     unsigned int flags)
+                     virConnectAuthPtr auth G_GNUC_UNUSED,
+                     virConfPtr conf G_GNUC_UNUSED, unsigned int flags)
 {
-    UNUSED(conn);
-    UNUSED(auth);
-    UNUSED(conf);
-    UNUSED(flags);
-    return 0;
+    virJailhouseDriverPtr driver = NULL;
+    uid_t uid = geteuid();
+
+    virCheckFlags(VIR_CONNECT_RO, VIR_DRV_OPEN_ERROR);
+
+    if (!virConnectValidateURIPath(conn->uri->path, "jailhouse", uid == 0))
+        return VIR_DRV_OPEN_ERROR;
+
+    if (!virFileExists(JAILHOUSE_CONFIG_FILE)) {
+        virReportError(VIR_ERR_INTERNAL_ERROR,
+                       _("Jailhouse configuration file %s does not exist"),
+                       JAILHOUSE_CONFIG_FILE);
+        return VIR_DRV_OPEN_ERROR;
+    }
+
+    if (access(JAILHOUSE_CONFIG_FILE, R_OK) < 0) {
+        virReportError(VIR_ERR_INTERNAL_ERROR,
+                       _
+                       ("Jailhouse configuration file %s is not accessible"),
+                       JAILHOUSE_CONFIG_FILE);
+        return VIR_DRV_OPEN_ERROR;
+    }
+
+    if (VIR_ALLOC(driver) < 0)
+        return VIR_DRV_OPEN_ERROR;
+
```

Comment:
How is this not leaked?

I think I mentioned in previous review, that what you want to do is to keep this @driver in conn->privateData. The reason is that virConnectOpen() returns a virConnect structure that is then passed to other APIs (e.g. virConnectListAllDomains() which in turn calls corresponding callback from driver table) and the driver's callback can get the pointer to the driver from conn->privateData. This concept is documented here:

https://libvirt.org/api.html

I also vaguely recall mentioning that the driver should be an object, so that it can be shared with other clients (e.g. two virsh-es from two consoles). I believe that we want all the clients to see the same state and not have one state per client. You can take a look at struct _vboxDriver. Its first member is "virObjectLockable parent". No need to get into details now (we can discuss on IRC later) but if you mimic that, then by using virObjectRef() and virObjectUnref() you can control the refcounter and thus share the object between two threads/clients.

Reply:
Fixed as per the latest commit.

I used the virMutex lock for locking the state of driver as the shared state would be used by the cell info list. 
Please check the latest amended commit.